plugin/pkg/admission/certificates/approval/admission.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package approval
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"k8s.io/klog"
+
+	"k8s.io/apiserver/pkg/admission"
+	genericadmissioninit "k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+
+	api "k8s.io/kubernetes/pkg/apis/certificates"
+)
+
+// PluginName is a string with the name of the plugin
+const PluginName = "CertificateApproval"
+
+// Register registers a plugin
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return newPlugin(), nil
+	})
+}
+
+// Plugin holds state for and implements the admission plugin.
+type Plugin struct {
+	*admission.Handler
+	authz authorizer.Authorizer
+}
+
+// SetAuthorizer sets the authorizer.
+func (p *Plugin) SetAuthorizer(authz authorizer.Authorizer) {
+	p.authz = authz
+}
+
+// ValidateInitialization ensures an authorizer is set.
+func (p *Plugin) ValidateInitialization() error {
+	if p.authz == nil {
+		return fmt.Errorf("%s requires an authorizer", PluginName)
+	}
+	return nil
+}
+
+var _ admission.ValidationInterface = &Plugin{}
+var _ genericadmissioninit.WantsAuthorizer = &Plugin{}
+
+// newPlugin creates a new CSR approval admission plugin
+func newPlugin() *Plugin {
+	return &Plugin{
+		Handler: admission.NewHandler(admission.Update),
+	}
+}
+
+// Validate verifies that the requesting user has permission to approve
+// CertificateSigningRequests for the specified signerName.
+func (p *Plugin) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	// Ignore all calls to anything other than 'certificatesigningrequests/approval'.
+	// Ignore all operations other than UPDATE.
+	if a.GetSubresource() != "approval" ||
+		a.GetResource().GroupResource() != api.Resource("certificatesigningrequests") ||
+		a.GetOperation() != admission.Update {
```

Comment:
How come? This only applies on UPDATE operations - CREATE should be ignored altogether/isn't a valid operation (unless that's why you say it shouldn't be needed 🙃)

Reply:
Since the plugin states it only handles `updates` via `Handler: admission.NewHandler(admission.Update)`, it should never be invoked in any other context:

https://github.com/kubernetes/kubernetes/blob/6858c25ee474ac2c90b49097c0ad04a61c6c54e1/staging/src/k8s.io/apiserver/pkg/admission/chain.go#L49-L51